### PR TITLE
Fix async checkpointing in nemo.lightning

### PR DESCRIPTION
**Pull request by hemildesai**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/lightning/pytorch/strategies/utils.py [hunk 1]
```diff
@@ -92,13 +92,10 @@ def ckpt_to_dir(filepath: Union[str, Path]) -> Path:
...
 
 
-def get_checkpoint_io(checkpoint_io, **kwargs):
-    if checkpoint_io is None:
-        checkpoint_io = MegatronCheckpointIO(**kwargs)
-        if kwargs.get("async_save", False):
-            checkpoint_io = AsyncFinalizableCheckpointIO(checkpoint_io)
-    elif isinstance(checkpoint_io, _WrappingCheckpointIO):
-        checkpoint_io.checkpoint_io = MegatronCheckpointIO()
+def get_checkpoint_io(**kwargs):
```
Comment (maanug-nv) [inline nemo/lightning/pytorch/strategies/utils.py:95]: I'd say this should now be renamed to create_checkpoint_io.

File: nemo/lightning/pytorch/strategies/megatron_strategy.py [hunk 2]
```diff
@@ -652,17 +652,19 @@ def load_model_state_dict(self, checkpoint: Mapping[str, Any], strict: bool = Tr
     @property
     @override
     def checkpoint_io(self) -> CheckpointIO:
-        return get_checkpoint_io(
-            self._checkpoint_io,
```
Comment (maanug-nv) [inline nemo/lightning/pytorch/strategies/megatron_strategy.py:656]: A bit confused why self._checkpoint_io reference is not being passed.

**Review by maanug-nv — DISMISSED** [review]

File: nemo/lightning/pytorch/strategies/megatron_strategy.py (hunk 2 above)
Comment (maanug-nv) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:656]: no, i meant before the PR, why doesn't `self._checkpoint_io` get set ? If it's passed by reference, it should be set ? We observed that it wasn't being set by `get_checkpoint_io`, but I don't understand why.

**Comment by maanug-nv** [comment]
can we add some test for async ckpt and make sure unfinished checkpoints are not present once finished?

**Comment by hemildesai** [comment]
> can we add some test for async ckpt and make sure unfinished checkpoints are not present once finished?

Yeah for sure, but I would like to address it in a separate PR since there's no easy way to add this test as of now.

File: nemo/lightning/pytorch/strategies/megatron_strategy.py (hunk 2 above)
Comment (maanug-nv) [reply nemo/lightning/pytorch/strategies/megatron_strategy.py:656]: ok, makes sense. good to know

**Review by maanug-nv — APPROVED** [review]
re-approving

**Review by blahBlahhhJ — APPROVED** [review]
LGTM
